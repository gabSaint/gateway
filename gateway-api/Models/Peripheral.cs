using gateway_api.Contexts;
using gateway_api.Extensions;
using gateway_api.Models.Validations;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace gateway_api.Models
{
    public class Peripheral
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public string Vendor { get; set; }

        [Required]
        public string Date { get; set; }

        [Required]
        [ValidStatus]
        public string Status { get; set; }

        [Required]
        [ForeignKey("Gateway")]
        public int GatewayId { get; set; } 

    }

    public static class PeripheralSeed
    {
        public static void InitData(DataContext context)
        {
            var rnd = new Random();

            var vendors = new[] { "ASL", "AFG", "HJP", "GER", "SPI" };

            var statuss = new[] { "online", "offline" };

            var dates = new[] { "Mon Jan 11 2021", "Tue Jan 12 2021", "Wed Jan 13 2021", "Thu Jan 14 2021", "Fri Jan 15 2021"  };

            context.Peripherals.AddRange(100.Times(x =>
            {
                var vendor = vendors[rnd.Next(0, 5)];
                var status = statuss[rnd.Next(0, 2)];
                var date = dates[rnd.Next(0, 5)];
                var id = $"{x,-3:000}";
                var gatewayId = rnd.Next(0, 20);

                return new Peripheral
                {
                    Id = x,
                    Vendor = vendor,
                    Date = date,
                    Status = status,
                    GatewayId = gatewayId
                };
            }));

            context.SaveChanges();
        }
    }
}
