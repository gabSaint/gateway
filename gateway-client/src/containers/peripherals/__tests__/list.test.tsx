import React from "react";
import { ReactWrapper, shallow, ShallowWrapper } from "enzyme";

import ListPeripherals, { Props as PeriferalProps } from "../list";
import Peripheral from "../../../models/peripheral";

describe("ListPeripherals", () => {
  let wrapperListPeripherals: ShallowWrapper | ReactWrapper;

  beforeEach(() => {
    wrapperListPeripherals = shallow(<ListPeripherals peripherals={[]} />);
  });

  describe("Table", () => {
    it("renders Table", () => {
      const table = wrapperListPeripherals.find("Table");
      expect(table).toHaveLength(1);
    });

    it("renders less than 11 table rows", () => {
      const tableRows = wrapperListPeripherals.find("tr");
      expect(tableRows.length).toBeLessThanOrEqual(11);
    });

    it("renders 5 td foreach tr in body", () => {
      const tableBodyRows = wrapperListPeripherals.find("tbody>tr");
      const tableBodycolumns = wrapperListPeripherals.find("tbody>tr>td");
      expect(tableBodycolumns.length).toBe(5 * tableBodyRows.length);
    });
  });

  describe("Buttons", () => {
    let tablebodyLastcolumns;

    beforeEach(() => {
      tablebodyLastcolumns = wrapperListPeripherals.find("td:last-child");
    });

    it("renders edit buttons for each peripheral", () => {
      const editButtons = wrapperListPeripherals.find(
        "[data-test='form-button-edit']"
      );
      expect(editButtons).toHaveLength(tablebodyLastcolumns.length);
    });
  });
});

describe("When a peripheral is passed to the component", () => {
  let wrapperListPeripherals: ReactWrapper | ShallowWrapper;
  let props: PeriferalProps;

  beforeEach(() => {
    props = {
      peripherals: [new Peripheral(1, "ASF", "21/23/24", "online", 1)],
    };
    wrapperListPeripherals = shallow(<ListPeripherals {...props} />);
  });

  it("displays the peripheral in the table", () => {
    const tableRowsColumns = wrapperListPeripherals.find("tr>td");
    expect(tableRowsColumns).toHaveLength(props.peripherals.length * 5);

    const attr = ["1", "ASF", "21/23/24", "online"];
    attr.forEach((at: string, i: number) => {
      expect(tableRowsColumns.at(i).text()).toBe(at);
    });
  });
});
