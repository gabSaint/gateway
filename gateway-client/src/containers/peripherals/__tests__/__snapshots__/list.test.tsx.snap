// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPeripherals renders correctly when listing a gateway's peripherals 1`] = `
Array [
  <h2>
    Peripherals
  </h2>,
  <table
    className="ui definition table"
  >
    <thead>
      <tr>
        <th>
          UID
        </th>
        <th>
          Vendor
        </th>
        <th>
          Date
        </th>
        <th>
          Status
        </th>
        <th>
          GatewayId
        </th>
        <th>
          Operations
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td>
          ASF
        </td>
        <td>
          21/23/24
        </td>
        <td>
          online
        </td>
        <td>
          1
        </td>
        <td>
          <div>
            <button
              className="button-edit"
              data-test="form-button-edit"
            >
              edit
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>,
]
`;

exports[`ListPeripherals renders correctly when listing all peripherals 1`] = `
Array [
  <h2>
    Peripherals
  </h2>,
  <table
    className="ui definition table"
  >
    <thead>
      <tr>
        <th>
          UID
        </th>
        <th>
          Vendor
        </th>
        <th>
          Date
        </th>
        <th>
          Status
        </th>
        <th>
          Operations
        </th>
      </tr>
    </thead>
    <tbody />
  </table>,
]
`;
